src/dstack/_internal/core/backends/gcp/resources.py

```diff
@@ -226,3 +228,32 @@ def generate_random_resource_name() -> str:
     return random.choice(string.ascii_lowercase) + "".join(
         random.choice(string.ascii_lowercase + string.digits) for _ in range(40)
     )
+
+
+def create_tpu_node_struct(
+    instance_name: str,
+    startup_script: str,
+    authorized_keys: List[str],
+    spot: bool,
+    labels: Dict[str, str],
+) -> tpu_v2.Node:
+    node = tpu_v2.Node()
+    if spot:
+        node.scheduling_config = tpu_v2.SchedulingConfig(preemptible=True)
+    node.accelerator_type = instance_name
+    node.runtime_version = "tpu-ubuntu2204-base"
+    node.network_config = tpu_v2.NetworkConfig(enable_external_ips=True)
+    ssh_keys = "\n".join(f"ubuntu:{key}" for key in authorized_keys)
+    node.metadata = {"ssh-keys": ssh_keys, "startup-script": startup_script}
+    node.labels = labels
+    return node
+
+
+def wait_for_operation(operation: Operation, verbose_name: str = "operation", timeout: int = 300):
```

Comment:
@Bihan, Any reason to define new wait_for_operation() vs using existing wait_for_extended_operation()? I suppose I mentioned wait_for_operation() in the review comment but I actually meant using existing wait_for_extended_operation().

As of now, wait_for_operation() seems incorrect since it will return result in case of an error, which won't be handled.

Reply:
@r4victor I defined a new wait_for_operation() because `self.tpu_client.create_node(request=create_node_request)` returns type `Operation`. The type `Operation` does not have field `error_code` like type `ExtendedOperation` used in wait_for_extended_operation(). 

To check the error handling I set default timeout to 3s and it resulted as below 
`[21:04:49] ERROR    dstack._internal.core.backends.gcp.resources:256 Error during tpu instance creation: Operation did not complete within the designated timeout
                    of 3 seconds.                                                                                                                                
           ERROR    dstack._internal.core.backends.gcp.resources:257 Operation ID: <google.api_core.operation.Operation object at 0x129af3710>    
 RuntimeError: Operation did not complete within the designated timeout of 3 seconds.                                                         
           DEBUG    dstack._internal.server.background.tasks.process_submitted_jobs:159 job(1e2ec0)witty-stingray-1-0-0: provisioning failed                     
           INFO     dstack._internal.server.background.tasks.process_runs:333 run(f96c86)witty-stingray-1: run status has changed SUBMITTED -> TERMINATING       
           INFO     dstack._internal.server.services.jobs:265 job(1e2ec0)witty-stingray-1-0-0: job status is FAILED, reason: FAILED_TO_START_DUE_TO_NO_CAPACITY  
[21:16:34] INFO     dstack._internal.server.services.runs:821 run(f96c86)witty-stingray-1: run status has changed TERMINATING -> FAILED, reason: JOB_FAILED   `

Any suggestions are welcome.

Reply:
Ah, I see now. Then it should be ok.